nexus/reconfigurator/planning/src/system.rs

```diff
@@ -0,0 +1,578 @@
+        let sled_subnets = Box::new(
+            rack_subnet
+                .subnets(SLED_PREFIX)
+                .unwrap()
+                .skip(1)
+                .map(|s| Ipv6Subnet::new(s.network())),
+        );
+
+        // Policy defaults
+        let target_nexus_zone_count = NEXUS_REDUNDANCY;
+        // IPs from TEST-NET-1 (RFC 5737)
+        let service_ip_pool_ranges = vec![IpRange::try_from((
+            "192.0.2.2".parse::<Ipv4Addr>().unwrap(),
+            "192.0.2.20".parse::<Ipv4Addr>().unwrap(),
+        ))
+        .unwrap()];
+
+        SystemDescription {
+            sleds: IndexMap::new(),
+            collector: None,
+            sled_subnets,
+            available_non_scrimlet_slots,
+            available_scrimlet_slots,
+            target_nexus_zone_count,
+            service_ip_pool_ranges,
+        }
+    }
+
+    /// Returns a complete system deployed on a single Sled
+    pub fn single_sled() -> anyhow::Result<Self> {
+        let mut builder = SystemDescription::new();
+        let sled = SledBuilder::new();
+        builder.sled(sled)?;
+        Ok(builder)
+    }
+
+    /// Returns a complete system resembling a full rack
+    pub fn full_rack() -> anyhow::Result<Self> {
+        let mut builder = SystemDescription::new();
+        for slot_number in 1..32 {
+            let mut sled = SledBuilder::new();
+            if slot_number == 14 || slot_number == 16 {
+                sled = sled.sled_role(SledRole::Scrimlet);
+            }
+            builder.sled(sled)?;
+        }
+        Ok(builder)
+    }
+
+    pub fn collector_label<S>(&mut self, collector_label: S) -> &mut Self
+    where
+        String: From<S>,
+    {
+        self.collector = Some(String::from(collector_label));
+        self
+    }
+
+    pub fn target_nexus_zone_count(&mut self, count: usize) -> &mut Self {
+        self.target_nexus_zone_count = count;
+        self
+    }
+
+    pub fn service_ip_pool_ranges(
+        &mut self,
+        ranges: Vec<IpRange>,
+    ) -> &mut Self {
+        self.service_ip_pool_ranges = ranges;
+        self
+    }
+
+    /// Add a sled to the system, as described by a SledBuilder
+    pub fn sled(&mut self, sled: SledBuilder) -> anyhow::Result<&mut Self> {
+        let sled_id = sled.id.unwrap_or_else(Uuid::new_v4);
+        ensure!(
+            !self.sleds.contains_key(&sled_id),
+            "attempted to add sled with the same id as an existing one: {}",
+            sled_id
+        );
+        let sled_subnet = self
+            .sled_subnets
+            .next()
+            .ok_or_else(|| anyhow!("ran out of IPv6 subnets for sleds"))?;
+        let hardware_slot = if let Some(slot) = sled.hardware_slot {
+            // If the caller specified a slot number, use that.
+            // Make sure it's still available, though.
+            if !self.available_scrimlet_slots.remove(&slot)
+                && !self.available_non_scrimlet_slots.remove(&slot)
```

Comment:
I don't think so?  If the slot is available, it would be in one map _or_ the other.  `remove()` returns whether it was there.  So this is saying: if it wasn't in the first map AND it wasn't in the second one, that's an error.

Reply:
Ahh sorry, I misunderstood this. Thanks.